binary_search_tree/tree.py

```diff
@@ -14,35 +14,146 @@ class Tree:
-        pass
+
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
+
+    #Iterative solution
+    #Time complexity: 0(log n) -  if balanced
+    #Space complexity: 0(1) - will always only increase by 1 if we add
+    # if self.root == None:
+    #     self.root = TreeNode(key, value=None)
+    # else:
+    #     previous = None
+    #     current = self.root
+
+    #     while current != None:
+    #         previous = current
+    #         if key <= current.key:
+    #             current = current.left
+    #         else:
+    #             current = current.right
+        
+    #     if key <= previous.key:
+    #         parent.left == TreeNode(key, value)
+    #     else:
+    #         parent.right = TreeNode(key, value)
+
 
     # Time Complexity: 
     # Space Complexity: 
     def find(self, key):
+
+        if self.root == None:
+            return None
+
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else:
+                current = current.left
+
+        return None
+
+
+    def inorder_helper(self):
         pass
 
     # Time Complexity: 
     # Space Complexity: 
     def inorder(self):
-        pass
+    #left, root, right
+        nodes_array = []
+        current =self.root
+
+        if self.root == None:
+            return []
 
+        while current != None:
+            nodes_array = self.inorder(current.left)
+            nodes_array.append(current.value)
+            nodes_array = nodes_array + self.inorder(current.right) 
+        return nodes_array 
+
+    def preorder_helper(self, current, nodes_array):
+        if current != None:
+            nodes_array.append({"key": current.key, "value": current.value})
+            self.preorder_helper(current.left, nodes_array)
+            self.preorder_helper(current.right, nodes_array)
+        
+        #or in a more recognizably recursive format 
+        #if current = none:
+        #return None
+        # nodes_array.append({"key": current.key, "value": current.value})
+        # self.preorder_helper(current.left, nodes_array)
+        # self.preorder_helper(current.right, nodes_array)
+
+        
+    
     # Time Complexity: 
-    # Space Complexity:     
+    # Space Complexity:  
     def preorder(self):
-        pass
+    #root, left, right
+        nodes_array = []
+        self.preorder_helper(self.root, nodes_array)
+        return nodes_array
+
 
     # Time Complexity: 
-    # Space Complexity:     
+    # Space Complexity:      
     def postorder(self):
+    #left, right, root
+        nodes_array = []
+        if self.root != None:
+            nodes_array.append(self.postorder(self.root.left))
+            nodes_array.append(self.postorder(self.root.right))
+        return nodes_array
```

Comment:
⚠️ This is also not working.  Take a look at the preorder method and see what you can do to adapt that method to this one.